web-api/src/personalization/routes.rs

```diff
@@ -465,12 +465,24 @@ impl UnvalidatedSemanticSearchQuery {
 
 #[derive(Deserialize)]
 struct UnvalidatedInputDocument {
-    id: String,
+    id: Option<String>,
+    query: Option<String>,
 }
 
 impl UnvalidatedInputDocument {
     fn validate(self) -> Result<InputDocument, Error> {
-        Ok(InputDocument::Ref(self.id.try_into()?))
+        if let (Some(_id), Some(_query)) = (&self.id, &self.query) {
+            Err(BadRequest::from(
+                "either id or query must be present in the request, but both were found",
+            )
+            .into())
+        } else if let Some(id) = self.id {
+            Ok(InputDocument::Ref(id.try_into()?))
+        } else if let Some(query) = self.query {
+            Ok(InputDocument::Query(query))
+        } else {
+            Err(BadRequest::from("either id or query must be present in the request").into())
+        }
```

Comment:
:art:  for this kind a if let matching a `match` is more appropriate, it has the nice side effect that it enforces at compiler time that you match on all combinations.

```suggestion
        match (self.id, self.query) {
            (Some(_), Some(_)) => Err(BadRequest::from(
                "either id or query must be present in the request, but both were found",
            )
            .into()),
            (None, Some(query)) => Ok(InputDocument::Query(query)),
            (Some(id), None) => Ok(InputDocument::Ref(id.try_into()?)),
            (None, None) => {
                Err(BadRequest::from("either id or query must be present in the request").into())
            }
        }
```

What the appropriate order is on the other hand is preference, some prefer a ordering based on the match patter, some would order the non error branches above the error branches and some would do that the other way around. Ordering based on the match pattern has the benefit that it can be applied in all contexts, so I often prefer that.

Reply:
Yup!
I'll change to your suggestion